connect/api/src/main/java/org/apache/kafka/connect/sink/SinkRecord.java

```diff
@@ -65,7 +180,8 @@ public SinkRecord newRecord(String topic, Integer kafkaPartition, Schema keySche
     @Override
     public SinkRecord newRecord(String topic, Integer kafkaPartition, Schema keySchema, Object key, Schema valueSchema, Object value,
                                 Long timestamp, Iterable<Header> headers) {
-        return new SinkRecord(topic, kafkaPartition, keySchema, key, valueSchema, value, kafkaOffset(), timestamp, timestampType, headers);
+        return new SinkRecord(topic, kafkaPartition, keySchema, key, valueSchema, value, kafkaOffset(), timestamp, timestampType, headers,
+                originalTopic(), originalKafkaPartition(), originalKafkaOffset());
```

Comment:
Nit: for consistency's sake, can we change this to use fields directly instead of access method?
```suggestion
        return new SinkRecord(topic, kafkaPartition, keySchema, key, valueSchema, value, kafkaOffset, timestamp, timestampType, headers,
                originalTopic, originalKafkaPartition, originalKafkaOffset);
```

Reply:
I figured this was intentional to differentiate between the use of instance fields and method arguments and chose to stick with the convention used here. Examples - [1](https://github.com/apache/kafka/blob/a1f6ab69387deb10988461152a0087f0cd2827c4/connect/api/src/main/java/org/apache/kafka/connect/sink/SinkRecord.java#L62), [2](https://github.com/apache/kafka/blob/a1f6ab69387deb10988461152a0087f0cd2827c4/connect/api/src/main/java/org/apache/kafka/connect/sink/SinkRecord.java#L68), [3](https://github.com/apache/kafka/blob/a1f6ab69387deb10988461152a0087f0cd2827c4/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/InternalSinkRecord.java#L54-L55)

Reply:
There's a bug in the [third example](https://github.com/apache/kafka/blob/a1f6ab69387deb10988461152a0087f0cd2827c4/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/InternalSinkRecord.java#L54-L55) 😱

The `newRecord` method accepts an `Iterable<Header> headers` parameter, but then when constructing the `InternalSinkRecord` that's used as the return value, it passes in `headers()` instead of `headers` as the parameter for the new record's headers, which causes the headers that were supplied to the `newRecord` method to be ignored.

It'd be great if we could fix that bug (and possibly add testing coverage for it). And IMO this is a decent reason to abandon using methods to differentiate between fields and parameters, though as long as there aren't any bugs in this PR I could live with it if you feel strongly that it's still worth making that distinction.

Reply:
Haha wow I completely glossed over that, nice catch.

> It'd be great if we could fix that bug (and possibly add testing coverage for it)

I think it might be worth filing a separate bug ticket and PR to fix that so that we could also backport it as well? We did discuss whether or not the method is even useful [here](https://github.com/apache/kafka/pull/14024#discussion_r1265694650), but given that we technically do support such use cases (and can't prove that it isn't used), it'd probably be good to backport a fix for such a bad bug if possible. WDYT?

> And IMO this is a decent reason to abandon using methods to differentiate between fields and parameters

Yeah, I'm definitely convinced since I missed it as well 😆

Reply:
A separate ticket/PR sounds great, thanks! Feel free to ping me on the PR if you implement the fx.